File: 322-CoinChange/322-CoinChange.py
# Last updated: 23/6/2025, 3:04:08 am
inf = float('inf')
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        coins.sort()
        @cache
        def rec(target):
            if target == 0:
                return 0
            
            ans = inf
            for coin in coins:
                if target-coin < 0: break
                ans = min(ans, rec(target - coin) + 1)
            return ans
        
        ans = rec(amount)
        return -1 if ans == float('inf') else ans

File: 322-CoinChange/322-CoinChange.cpp
// Last updated: 12/6/2025, 5:50:38 am
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        long dp[amount+1];
        for (int i = 1; i <= amount; ++i){
            dp[i] = INT_MAX-1;
        }
        dp[0] = 0;
        for (long i=0; i<=amount; i++){
            for (const long &coin: coins){
                if (i+coin <= amount){
                    dp[i+coin] = min(dp[i]+1, dp[i+coin]);
                }
            }
        }
        return dp[amount] == INT_MAX-1 ? -1 : dp[amount];
    }
};